### PR TITLE
Classes and Structures for ARP packet counters

### PR DESCRIPTION
**Pull request by FrancescoStaccini**
Hash table of ARP packet counters, that represent the number of packet sent/received from/to specific devices.

(not finished yet)

### CODE REVIEW

File: src/NetworkInterface.cpp [hunk 1]
```diff
@@ -3173,6 +3199,14 @@ struct mac_find_info {
 
 /* **************************************************** */
 
+struct arp_stats_element_find_info {
+  u_int8_t src_mac[6];
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp:3247]: never used?

File: src/NetworkInterface.cpp [hunk 2]
```diff
@@ -3173,6 +3199,14 @@ struct mac_find_info {
...
+  u_int8_t dst_mac[6];
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp:3248]: never used?

File: src/NetworkInterface.cpp [hunk 3]
```diff
@@ -3229,6 +3263,25 @@ static bool find_mac_by_name(GenericHashEntry *h, void *user_data, bool *matched
...
 /* **************************************************** */
 
+
+static bool find_arp_stats_element_by_macs_name(GenericHashEntry *h, void *user_data, bool *matched) {
+
+  struct arp_stats_element_find_info *info = (struct arp_stats_element_find_info*)user_data;
+
+  ArpStatsMatrixElement *e = (ArpStatsMatrixElement*)h;
+
+  if( (info->elem == NULL) && !info->elem->equal(e->getSourceMac(), e->getDestinationMac()) ){
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp]: code inside this is IF is never reached and is going to crash with a SEGFAULT when info->elem is null

File: src/NetworkInterface.cpp [hunk 4]
```diff
@@ -5364,6 +5427,61 @@ Mac* NetworkInterface::getMac(u_int8_t _mac[6], bool createIfNotPresent) {
...
+//remember that if src and dst mac are reversed, you must reverse ReqReplyStats from ArpStats
+bool NetworkInterface::getArpStatsMatrixElementInfo(lua_State* vm, char *src_mac, char *dst_mac) {
+  struct arp_stats_element_find_info info;
+  bool ret;
+  u_int32_t begin_slot = 0;
+  bool walk_all = true;
+
+  memset(&info, 0, sizeof(info));
+
+  disablePurge(false);
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp]: enable / disable purge not necessary

File: src/NetworkInterface.cpp [hunk 5]
```diff
@@ -5364,6 +5427,61 @@ Mac* NetworkInterface::getMac(u_int8_t _mac[6], bool createIfNotPresent) {
...
+  ArpStatsMatrixElement *ret = NULL;
+
+  if ( _src_mac == NULL || _dst_mac == NULL) return NULL;
+
+  ret = arp_hash_matrix->get(_src_mac, _dst_mac);
+  
+  if ( ret == NULL && createIfNotPresent ){
+    try{ 
+      if( (ret = new ArpStatsMatrixElement(this, _src_mac, _dst_mac)) != NULL)
+        arp_hash_matrix->add(ret);
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp]: not handled the case when this add fails: must delete and free ret

File: src/NetworkInterface.cpp [hunk 6]
```diff
@@ -2461,17 +2476,27 @@ bool NetworkInterface::dissectPacket(u_int32_t bridge_iface_idx,
...
 
+
+  ArpStatsMatrixElement* e;
 	if(arp_opcode == 0x1 /* ARP request */) {
 	  arp_requests++;
 	  srcMac->incSentArpRequests();
 	  dstMac->incRcvdArpRequests();
+
+    e = getArpHashMatrixElement(srcMac->get_mac(), dstMac->get_mac(), true);
+    e->AddOneSentRequests();
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp]: please use the ntopng convention and name this kind of methods similar to incSentArpRequests(), incRcvdArpRequuests() etc..

File: src/ArpStatsMatrixElement.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  return(rc);
+}
+
+/* *************************************** */
+
+bool ArpStatsMatrixElement::equal(const u_int8_t _src_mac[6], const u_int8_t _dst_mac[6]) {
+    if(! _src_mac || !_dst_mac)
+        return false;
+
+    if(  memcmp(src_mac, _src_mac, 6) == 0 && memcmp(dst_mac, _dst_mac, 6) == 0 )
```
Comment (simonemainardi) [inline src/ArpStatsMatrixElement.cpp]: please remove unnecessary spaces

File: src/NetworkInterface.cpp [hunk 8]
```diff
@@ -164,6 +164,9 @@ NetworkInterface::NetworkInterface(const char *name,
 
     macs_hash = new MacHash(this, num_hashes, ntop->getPrefs()->get_max_num_hosts());
 
+    arp_hash_matrix = new ArpStatsHashMatrix(this, num_hashes,
+            ntop->getPrefs()->get_max_num_hosts());
```
Comment (simonemainardi) [inline src/NetworkInterface.cpp:168]: this maximum may not be enough as this is a matrix

**Review by simonemainardi — CHANGES_REQUESTED** [review]

**Comment by FrancescoStaccini** [comment]
thanks a lot @simonemainardi  for the review. I will apply the changes and follow the suggestions.

**Comment by simonemainardi** [comment]
@FrancescoStaccini are you having troubles in pushing the new version with the suggested changes? I'can see it, it's seems the commit from 1 day ago involves a pull request from 2015...

image

**Comment by FrancescoStaccini** [comment]
I hope I did not messed up, i was trying to sync my fork with ntop

**Comment by lucaderi** [comment]
Cleanup and restart

**Pull request closed without merging**